lib/models/addon-discovery.js

```diff
@@ -20,241 +20,237 @@ var findup     = require('findup');
-  var discovery = this;
-  var addons = Object.keys(this.dependencies(pkg, excludeDevDeps)).map(function(name) {
-    if (name !== 'ember-cli') {
-      var addonPath = this.resolvePackage(root, name);
-
-      if (addonPath) {
-        return discovery.discoverAtPath(addonPath);
-      }
-
-      // this supports packages that do not have a valid entry point
-      // script (aka `main` entry in `package.json` or `index.js`)
-      addonPath = path.join(nodeModulesPath, name);
-      var addon = discovery.discoverAtPath(addonPath);
-      if (addon) {
-        var chalk = require('chalk');
-
-        discovery.ui.writeLine(chalk.yellow('The package `' + name + '` is not a properly formatted package, we have used a fallback lookup to resolve it at `' + addonPath + '`. This is generally caused by an addon not having a `main` entry point (or `index.js`).'), 'WARNING');
-
-        return addon;
+    var inRepoAddons = this.discoverInRepoAddons(project.root, project.pkg);
+    var addons = projectAsAddon.concat(cliAddons, internalAddons, dependencyAddons, inRepoAddons);
+
+    return addons;
+  },
+
+  /**
+    This is one of the primary APIs for this class and is called by addons.
+    It returns a tree of plain objects that each contain information about a
+    discovered addon. Each node has `name`, `path`, `pkg` and
+    `childAddons` properties. The latter is an array containing any addons
+    discovered from applying the discovery process to that addon.
+
+    @private
+    @method discoverProjectAddons
+  */
+  discoverChildAddons: function(addon) {
+    debug('discoverChildAddons: %s(%s)', addon.name, addon.root);
+    var dependencyAddons = this.discoverFromDependencies(addon.root, addon.nodeModulesPath, addon.pkg, true);
+    var inRepoAddons = this.discoverInRepoAddons(addon.root, addon.pkg);
+    var addons = dependencyAddons.concat(inRepoAddons);
+
+    addons = addons.filter(function(childAddon) {
+      return !addon.shouldIncludeChildAddon || addon.shouldIncludeChildAddon(childAddon);
+    });
+
+    return addons;
+  },
+
+  /**
+    Returns an array containing zero or one nodes, depending on whether or not
+    the passed project is an addon.
+
+    @private
+    @method discoverFromProjectItself
+   */
+  discoverFromProjectItself: function(project) {
+    if (project.isEmberCLIAddon()) {
+      var addonPkg = this.discoverAtPath(project.root);
+      if (addonPkg) {
+        return [addonPkg];
       }
     }
-  }, this).filter(Boolean);
-  return addons;
-};
+    return [];
+  },
+
+  /**
+    Returns a tree based on the addons referenced in the provided `pkg` through
+    the package.json `dependencies` and optionally `devDependencies` collections,
+    as well as those discovered addons' child addons.
+
+    @private
+    @method discoverFromDependencies
+   */
+  discoverFromDependencies: function(root, nodeModulesPath, pkg, excludeDevDeps) {
+    var discovery = this;
```

Comment:
preexisting https://github.com/ember-cli/ember-cli/blob/master/lib/models/addon-discovery.js#L107-L108

should we deal with this here or another PR?